freeipa/src/main/java/com/sequenceiq/freeipa/service/freeipa/user/UserSyncService.java

```diff
@@ -117,22 +121,37 @@ public Operation synchronizeUsers(String accountId, String actorCrn, Set<String>
         LOGGER.info("Starting operation [{}] with status [{}]", operation.getOperationId(), operation.getStatus());
 
         if (operation.getStatus() == OperationState.RUNNING) {
-            boolean fullSync = userCrnFilter.isEmpty() && machineUserCrnFilter.isEmpty();
-            if (fullSync) {
-                long currentTime = Instant.now().toEpochMilli();
-                stacks.forEach(stack -> {
-                    UserSyncStatus userSyncStatus = userSyncStatusService.getOrCreateForStack(stack);
-                    userSyncStatus.setLastFullSyncStartTime(currentTime);
-                    userSyncStatusService.save(userSyncStatus);
+            try {
+                ThreadBasedUserCrnProvider.doAs(INTERNAL_USER_CRN, () -> {
+                    boolean fullSync = userCrnFilter.isEmpty() && machineUserCrnFilter.isEmpty();
+                    if (fullSync) {
+                        long currentTime = Instant.now().toEpochMilli();
+                        stacks.forEach(stack -> {
+                            UserSyncStatus userSyncStatus = userSyncStatusService.getOrCreateForStack(stack);
+                            userSyncStatus.setLastFullSyncStartTime(currentTime);
+                            userSyncStatusService.save(userSyncStatus);
+                        });
+                    }
+                    asyncSynchronizeUsers(operation.getOperationId(), accountId, actorCrn, stacks, userCrnFilter, machineUserCrnFilter, fullSync);
                 });
+            } catch (Throwable t) {
+                try {
+                    LOGGER.error("Failed to start operation [{}]. Attempting to mark failure in database then re-throwing", operation, t);
+                    operationService.failOperation(accountId, operation.getOperationId(),
+                            "Failed to start operation: " + t.getLocalizedMessage());
+                } catch (Exception e) {
+                    LOGGER.error("Failed to mark operation [{}] as failed in database.", operation, e);
+                } finally {
+                    throw t;
+                }
```

Comment:
this looks almost the same like in `internalSynchronizeUsers`, could you move these into a separate method?

Reply:
done